otherlibs/systhreads/st_stubs.c

```diff
@@ -197,22 +199,31 @@ void caml_thread_restore_runtime_state(void)
 
 static void caml_thread_enter_blocking_section(void)
 {
-  /* Save the current runtime state in the thread descriptor
-     of the current thread */
-  caml_thread_save_runtime_state();
-  /* Tell other threads that the runtime is free */
-  st_masterlock_release(&Thread_main_lock);
+  if (This_thread != NULL) {
+    /* Save the current runtime state in the thread descriptor
+       of the current thread */
+    caml_thread_save_runtime_state();
+    /* Tell other threads that the runtime is free */
+    st_masterlock_release(&Thread_main_lock);
+  } else {
+    caml_enter_blocking_section_default();
+  }
 }
 
 static void caml_thread_leave_blocking_section(void)
 {
-  /* Wait until the runtime is free */
-  st_masterlock_acquire(&Thread_main_lock);
-  /* Update Active_thread to point to the thread descriptor corresponding to
-     the thread currently executing */
-  Active_thread = st_tls_get(caml_thread_key);
-  /* Restore the runtime state from the curr_thread descriptor */
-  caml_thread_restore_runtime_state();
+  caml_thread_t this_thread = This_thread;
+  if (this_thread != NULL) {
+    /* Wait until the runtime is free */
+    st_masterlock_acquire(&Thread_main_lock);
+    /* Update Active_thread to point to the thread descriptor corresponding to
+       the thread currently executing */
+    Active_thread = this_thread;
+    /* Restore the runtime state from the curr_thread descriptor */
+    caml_thread_restore_runtime_state();
+  } else {
+    caml_leave_blocking_section_default();
```

Comment:
My intuition is that `caml_thread_leave_blocking_section` and its (enter counterpart) would not be called if initialization isn't done for `Thread`.
(as these are set as callback only if `caml_thread_initialize` is called)
Is this extra handling required?

Reply:
Yes, for domain ≠ 0, there is a gap between spawning and threads init during which arbitrary OCaml code can run. The solution is not really nice.

I do not really like this solution (look at the removal of `[@noalloc]` on Thread.self). It would be simpler to avoid this by calling the domain threads init with a hook on the C side. Would you or someone else would be willing to look into such an alternative solution?

Reply:
To be quite frank I think the solution here works and we should merge the branch as is.

I agree this is not quite elegant, and I've never been a huge fan of the initialisation process for threads to be called from OCaml.

Maybe we should just open an issue and address this later?

Reply:
> we should merge the branch as is

it needs an approval first :-)